File: version_finder.py
# version_finder.py

import subprocess
import sys
import os

class VersionFinder:
    """
    A class for finding version information in a Git repository.

    Args:
        path (str, optional): The path to the Git repository. If not provided, the current working directory is used.

    Attributes:
        repository_path (str): The absolute path to the Git repository.
        submodules (list): A list of submodules in the main project.
        branches (list): A list of all branches in the main project.

    Methods:
        get_submodules(): Get the list of submodules in the repository.
        get_branches(): Get the list of branches in the repository.
        is_valid_branch(branch): Check if a branch is valid.
        is_valid_submodule(submodule): Check if a submodule is valid.
        is_valid_commit_sha(commit_sha, branch, submodule): Check if a commit SHA is valid for a given branch and submodule.
        get_sha_of_first_commit_including_target(target, branch, submodule): Get the SHA of the first commit that includes a target commit in a given branch and submodule.
        get_all_logs_until_commit(commit, branch, submodule): Get all the logs from the current commit until a specific commit in a given branch and submodule.
        find_first_commit_with_version(commit, branch, submodule): Find the first commit with a version in the logs from the current commit until a specific commit in a given branch and submodule.
    """

    def __init__(self, path=None) -> None:
        """
        Initialize the VersionFinder object.

        Args:
            path (str, optional): The path to the Git repository. If not provided, the current working directory is used.
        """
        # Set CWD to the repository path
        if not path:
            self.repository_path = os.getcwd()
        else:
            self.repository_path = os.path.abspath(path)

        # Make sure the path is a valid git repository
        try:
            subprocess.check_output(["git", "rev-parse"], cwd=self.repository_path, stderr=subprocess.DEVNULL)
        except subprocess.CalledProcessError:
            print(f"Invalid git repository path: {self.repository_path}")
            sys.exit(1)

        if not self.__is_clean_git_repo():
            print("The repository is not clean. Please commit or stash your changes.")
            sys.exit(1)
        # Get a list of submodules in the main project
        try:
            output = subprocess.check_output(["git", "submodule", "status"], cwd=self.repository_path, stderr=subprocess.DEVNULL)
            self.submodules = [line.split()[1] for line in output.decode("utf-8").splitlines()]
        except subprocess.CalledProcessError:
            print("Error fetching submodule information.")
            sys.exit(1)

        # Get a list of all branches in the main project
        try:
            output = subprocess.check_output(["git", "branch", "-r"], cwd=self.repository_path, stderr=subprocess.DEVNULL)
            self.branches = [line.strip() for line in output.decode("utf-8").splitlines()]
            # Remove the `HEAD -> <current_branch>` branch
            self.branches = self.branches[1:]
            # Remove the origin/ prefix
            self.branches = [branch.split("/", 1)[1] for branch in self.branches]
        except subprocess.CalledProcessError:
            print("Error fetching branch information.")
            sys.exit(1)

    def get_submodules(self):
        """
        Get the list of submodules in the repository.

        Returns:
            list: A list of submodules in the repository.
        """
        return self.submodules

    def get_branches(self):
        """
        Get the list of branches in the repository.

        Returns:
            list: A list of branches in the repository.
        """
        return self.branches

    def is_valid_branch(self, branch):
        """
        Check if a branch is valid.

        Args:
            branch (str): The branch name.

        Returns:
            bool: True if the branch is valid, False otherwise.
        """
        return branch in self.branches

    def is_valid_submodule(self, submodule):
        """
        Check if a submodule is valid.

        Args:
            submodule (str): The submodule path.

        Returns:
            bool: True if the submodule is valid, False otherwise.
        """
        return submodule in self.submodules

    def is_valid_commit_sha(self, commit_sha, branch, submodule=None):
        """
        Check if a commit SHA is valid for a given branch and submodule.

        Args:
            commit_sha (str): The commit SHA.
            branch (str): The branch name.
            submodule (str, optional): The submodule path. Defaults to None.

        Returns:
            bool: True if the commit SHA is valid, False otherwise.
        """
        try:
            # Checkout the branch
            subprocess.check_output(["git", "checkout", branch], cwd=self.repository_path, stderr=subprocess.DEVNULL)
            # Pull
            subprocess.check_output(["git", "pull"], cwd=self.repository_path, stderr=subprocess.DEVNULL)
            # Update submodules
            subprocess.check_output(["git", "submodule", "update", "--init"], cwd=self.repository_path, stderr=subprocess.DEVNULL)
            if submodule:
                # Go to the submodule directory
                path = os.path.join(self.repository_path, submodule)
            else:
                path = self.repository_path

            # Verify the commit exists in the submodule
            subprocess.check_output(["git", "show", commit_sha], cwd=path, stderr=subprocess.DEVNULL)
            return True
        except subprocess.CalledProcessError:
            return False

    def __is_clean_git_repo(self):
        '''
        Check if the repository is clean.

        Returns:
            bool: True if the repository is clean, False otherwise.
        '''
        try:
            subprocess.check_output(["git", "diff", "--quiet"], cwd=self.repository_path, stderr=subprocess.DEVNULL)
            return True
        except subprocess.CalledProcessError:
            return False


    def __is_ancestor(self, ancestor, commit, submodule=None):
        '''
        Check if the commit is an ancestor of the ancestor commit.

        Note: Assumes the two commits exists in the repository/submodule, and that we are the root of the repository/submodule.

        Args:
            ancestor (str): The ancestor commit SHA.
            commit (str): The commit SHA.
            submodule (str, optional): The submodule path. Defaults to None.

        Returns:
            bool: True if the commit is an ancestor, False otherwise.
        '''
        if submodule:
            path = os.path.join(self.repository_path, submodule)
        else:
            path = self.repository_path
        try:
            subprocess.check_output(["git", "merge-base", "--is-ancestor",ancestor, commit], cwd=path, stderr=subprocess.DEVNULL)
            return True
        except subprocess.CalledProcessError:
            return False

    def __get_pointer_to_submodule(self, submodule, branch=""):
        try:
            if branch:
                # Checkout the branch
                subprocess.check_output(["git", "checkout", branch], cwd=self.repository_path, stderr=subprocess.DEVNULL)
                # Update submodules
                subprocess.check_output(["git", "submodule", "update", "--init"], cwd=self.repository_path, stderr=subprocess.DEVNULL)

            # Get the commit SHA of the repository/submodule in the branch
            if submodule:
                output = subprocess.check_output(["git", "submodule", "status", submodule], cwd=self.repository_path, stderr=subprocess.DEVNULL)
                return output.decode("utf-8").split()[0]
            else:
                output = subprocess.check_output(["git", "rev-parse", "HEAD"], cwd=self.repository_path, stderr=subprocess.DEVNULL)
                return output.decode("utf-8").strip()
        except subprocess.CalledProcessError:
            return None

    def __get_submodule_pointer_at_specific_repo_commit(self, submodule, commit):
        try:
            output = subprocess.check_output(["git", "ls-tree", commit, submodule], cwd=self.repository_path, stderr=subprocess.DEVNULL)
            return output.decode("utf-8").split()[2]
        except subprocess.CalledProcessError:
            return None

    def get_sha_of_first_commit_including_target(self, target, branch, submodule=None):
        """
        Get the SHA of the first commit that includes a target commit in a given branch and submodule.

        Args:
            target (str): The target commit SHA.
            branch (str): The branch name.
            submodule (str, optional): The submodule path. Defaults to None.

        Returns:
            str: The SHA of the first commit that includes the target commit, or None if not found.
        """
        try:
            # Checkout the branch
            subprocess.check_output(["git", "checkout", branch], cwd=self.repository_path, stderr=subprocess.DEVNULL)
            # Update submodules
            subprocess.check_output(["git", "submodule", "update", "--init"], cwd=self.repository_path, stderr=subprocess.DEVNULL)
            if submodule:
                # Iterate over all commits in repository that changed the submodule
                output = subprocess.check_output(["git", "rev-list", "HEAD", "--", submodule], cwd=self.repository_path, stderr=subprocess.DEVNULL)
                commits_list = output.decode("utf-8").splitlines()
                for indx,commit in enumerate(commits_list):
                    submodule_commit = self.__get_submodule_pointer_at_specific_repo_commit(submodule, commit)
                    if not self.__is_ancestor(target, submodule_commit, submodule=submodule):
                        return commits_list[indx-1]
            else:
                return target
        except subprocess.CalledProcessError:
            return None

    def get_all_logs_until_commit(self, commit, branch, submodule=None):
        """
        Get all the logs from the current commit until a specific commit in a given branch and submodule.

        Args:
            commit (str): The commit SHA.
            branch (str): The branch name.
            submodule (str, optional): The submodule path. Defaults to None.

        Returns:
            str: The logs from the current commit until the specific commit.
        """
        try:
            # Checkout the branch
            subprocess.check_output(["git", "checkout", branch], cwd=self.repository_path, stderr=subprocess.DEVNULL)
            # Update submodules
            subprocess.check_output(["git", "submodule", "update", "--init"], cwd=self.repository_path, stderr=subprocess.DEVNULL)
            if submodule:
                path = os.path.join(self.repository_path, submodule)
            else:
                path = self.repository_path
            # Show all logs until the commit
            output = subprocess.check_output(["git", "log", f"{commit}..HEAD", "--oneline", "--reverse"], cwd=self.repository_path, stderr=subprocess.DEVNULL)
            return (output.decode("utf-8"))
        except subprocess.CalledProcessError:
            print("Error showing logs.")
            sys.exit(1)

    def find_first_commit_with_version(self, commit, branch, submodule=None):
        """
        Find the first commit with a version in the logs from the current commit until a specific commit in a given branch and submodule.

        Args:
            commit (str): The commit SHA.
            branch (str): The branch name.
            submodule (str, optional): The submodule path. Defaults to None.

        Returns:
            str: The SHA of the first commit with a version, or None if not found.
        """
        gitlog = self.get_all_logs_until_commit(commit, branch, submodule)
        for line in gitlog.splitlines():
            if "Version:" in line:
                print ("The first version including the commit is:")
                print (line)
                return line.split()[0]
        print ("No version found in the logs.")


if __name__ == "__main__":

    # Get the path of the repository from the command line invoke as argument
    if len(sys.argv) != 2:
        print("Usage: python version_finder.py [path_to_repository=.]")
        print("Setting repository path to the current directory.")
        path = None
    else:
        path = sys.argv[1]

    version_finder = VersionFinder(path)
    submodules = version_finder.get_submodules()
    branches = version_finder.get_branches()

    # Set default values
    selected_submodule = None
    selected_branch = "main"

    if submodules:
        print("Available submodules:")
        for submodule in submodules:
            print(f"- {submodule}")

        selected_submodule = input("Enter the submodule path (e.g., sub-module-A): ")
        # Verify the selected submodule
        if not version_finder.is_valid_submodule(selected_submodule):
            print("Invalid submodule path.")
            sys.exit(1)
    else:
        print("No submodules found.")

    print("\nAvailable branches:")
    for branch in branches:
        print(f"- {branch}")

    selected_branch = input("\nEnter the branch name: ")
    # Verify the selected branch
    if not version_finder.is_valid_branch(selected_branch):
        print("Invalid branch name.")
        sys.exit(1)

    selected_sha = input("Enter the commit SHA: ")

    if not version_finder.is_valid_commit_sha(selected_sha, selected_branch, selected_submodule):
        print("Invalid commit SHA.")
        sys.exit(1)

    first_commit_sha = version_finder.get_sha_of_first_commit_including_target(selected_sha, selected_branch, selected_submodule)
    print(f"The SHA of the first commit including the target is: {first_commit_sha}")

    print("\nLogs until the first commit including the target: (From Head to the first commit including the target)")
    print(version_finder.get_all_logs_until_commit(first_commit_sha, selected_branch, selected_submodule))

    version_finder.find_first_commit_with_version(first_commit_sha, selected_branch, selected_submodule)

File: version_finder.js
// version_finder.js

const gitP = require("simple-git");
const path = require("path");
const { exec } = require("child_process");
const util = require("util");
const execPromise = util.promisify(exec);
const fs = require("fs");

/**
 * Represents a VersionFinder object that provides methods to interact with a git repository.
 */
/**
 * Represents a VersionFinder object that is used to find versions in a Git repository.
 */
class VersionFinder {
  /**
   * Creates a new VersionFinder object.
   * @param {string} repositoryPath - The path to the git repository. Defaults to the current working directory.
   */
  constructor(repositoryPath = process.cwd()) {
    this.repositoryPath = path.resolve(repositoryPath);

    if (!fs.existsSync(this.repositoryPath)) {
      console.error(
        `The provided path "${this.repositoryPath}" does not exist.`
      );
      // Handle the error appropriately. For example, you might want to throw an error or set an internal state.
      throw new Error(
        `The provided path "${this.repositoryPath}" does not exist.`
      );
    }
    this.git = gitP(this.repositoryPath);
    this.submodules = [];
    this.branches = [];
    this.searchPatternRegex = null;
    this.isInitialized = false;
    this.hasChanges = false;
    this.snapshot = null;
  }

  async destructor() {
    console.log("In destructor");
    await this.restoreRepoSnapshot();
    this.repositoryPath = null;
    this.git = null;
    this.submodules = null;
    this.branches = null;
    this.searchPatternRegex = null;
    this.isInitialized = null;
    this.hasChanges = null;
  }

  /**
   * Sets the search pattern for the version finder.
   * @param {string} searchPattern - The search pattern to set.
   * @throws {Error} If the search pattern is not a string or is not a valid regex.
   */
  setSearchPattern(searchPattern) {
    // Validate the input is a string and not a regex
    if (typeof searchPattern !== "string") {
      throw new Error("The search pattern must be a string.");
    }

    // Validate the string is a valid regex
    if (
      !searchPattern.startsWith("/") ||
      searchPattern.lastIndexOf("/") === 0
    ) {
      throw new Error("The search pattern must be a valid regex.");
    }

    console.log("searchPattern: ", searchPattern);
    const regexBody = searchPattern.slice(1, searchPattern.lastIndexOf("/"));

    // Extract flags if present
    const regexFlags = searchPattern.slice(searchPattern.lastIndexOf("/") + 1);

    this.searchPatternRegex = new RegExp(regexBody, regexFlags);
    console.log("searchPatternRegex: ", this.searchPatternRegex);
  }

  /**
   * Checks if a Git repository is dirty (has uncommitted changes).
   * @param {SimpleGit} gitRepo - The SimpleGit instance representing the Git repository.
   * @returns {Promise<boolean>} - A Promise that resolves to `true` if the repository is dirty, or `false` otherwise.
   */
  async isRepoDirty(gitRepo) {
    const status = await gitRepo.status(["-uno"]);
    let isDirty = false;
    if (status.files.length > 0) {
      console.warn(
        "Warning: The repository has uncommitted changes. This may affect the results."
      );
      isDirty = true;
    }
    return isDirty;
  }

  /**
   * Initializes the VersionFinder object by checking if the repository is valid and fetching submodule and branch information.
   * @returns {Promise<Error|null>} - A promise that resolves to null if initialization is successful, or an Error object if there is an error.
   */
  async init() {
    if (this.isInitialized) {
      return;
    }
    await this.git.checkIsRepo().then((isRepo) => {
      if (!isRepo) {
        throw new Error(
          `The given path ${this.repositoryPath} is Not a git repository`
        );
      }
    });

    try {
      const submodules_raw = await this.git.subModule(["status"]);
      if (submodules_raw) {
        const submodules_raw_lines = submodules_raw.split("\n");
        if (submodules_raw_lines[submodules_raw_lines.length - 1] === "") {
          submodules_raw_lines.pop();
        }
        this.submodules = submodules_raw_lines.map(
          (line) => line.split(" ")[2]
        );
      } else {
        this.submodules = [];
      }
    } catch (e) {
      console.error("Error fetching submodule information.");
      this.submodules = [];
      throw e;
    }

    try {
      let branches = await this.git.branch();
      this.branches = branches.all.map((branch) =>
        branch.replace("origin/", "").replace("remotes/", "")
      );
      // Remove duplicates
      this.branches = [...new Set(this.branches)];
    } catch (e) {
      console.error("Error fetching branch information.");
      throw e;
    }

    // Check if repository is "clean" (no uncommitted changes)
    this.hasChanges = await this.isRepoDirty(this.git);

    // Save the repo status
    await this.saveRepoSnapshot(false);

    // Set the initialized flag
    this.isInitialized = true;
  }

  /**
   * Gets the list of submodules in the repository.
   * @returns {string[]} - An array of submodule names.
   */
  getSubmodules() {
    if (!this.isInitialized) {
      throw new Error("VersionFinder is not initialized.");
    }
    return this.submodules;
  }

  /**
   * Gets the list of branches in the repository.
   * @returns {string[]} - An array of branch names.
   */
  getBranches() {
    if (!this.isInitialized) {
      throw new Error("VersionFinder is not initialized.");
    }
    return this.branches;
  }

  /**
   * Checks if a branch is valid.
   * @param {string} branch - The branch name to check.
   * @returns {boolean} - true if the branch is valid, false otherwise.
   */
  isValidBranch(branch) {
    if (!this.isInitialized) {
      throw new Error("VersionFinder is not initialized.");
    }
    return this.branches.includes(branch);
  }

  /**
   * Checks if a submodule is valid.
   * @param {string} submodule - The submodule name to check.
   * @returns {boolean} - true if the submodule is valid, false otherwise.
   */
  isValidSubmodule(submodule) {
    if (!this.isInitialized) {
      throw new Error("VersionFinder is not initialized.");
    }
    return this.submodules.includes(submodule);
  }

  /**
   * Takes a snapshot of the given git repository.
   *
   * @param {GitRepo} gitRepo - The git repository object.
   * @returns {Promise<Object>} A promise that resolves to an object containing the snapshot information.
   */
  async takeRepoSnapshot(gitRepo, allow_to_stash = true) {
    const snapshot = {
      commitHash: null,
      branch: null,
      stashId: null,
    };
    if ((await this.isRepoDirty(gitRepo)) && allow_to_stash) {
      // Stash the changes
      const stash = await gitRepo.stash(["save", "VersionFinder snapshot"]);
      console.log("stash: ", stash);
      // Get the id by fetching the list of stashes
      const stashList = await gitRepo.stash(["list"]);
      for (const stashEntry of stashList.split("\n")) {
        console.log("stashEntry: ", stashEntry);
        if (stashEntry.includes("VersionFinder snapshot")) {
          snapshot.stashId = stashEntry.match(/stash@\{0\}/)[0];
          break;
        }
      }
    } else {
      snapshot.stashId = null;
    }
    let branch = await gitRepo.revparse(["--abbrev-ref", "HEAD"]);
    if (branch === "HEAD") {
      branch = null;
    }
    const commitHash = await gitRepo.revparse(["HEAD"]);
    snapshot.commitHash = commitHash;
    snapshot.branch = branch;

    return snapshot;
  }

  /**
   * Saves the current state of the repository to a snapshot.
   *
   * @throws {Error} If VersionFinder is not initialized.
   */
  async saveRepoSnapshot(allow_to_stash = true) {
    console.log("In saveRepoSnapshot");
    if (this.snapshot && this.snapshot.isAllowedToStash === true) {
      console.log("In saveRepoSnapshot: snapshot already exists");
      return;
    }

    // Create a snapshot object
    const snapshot = {
      isAllowedToStash: allow_to_stash,
      mainRepo: {
        commitHash: "",
        branch: "",
        stashId: "",
      },
      submodules: {},
    };

    // Iterate over all submodules
    console.log("this.submodules: ", this.submodules);
    for (const submodule of this.submodules) {
      const submodulePath = path.join(this.repositoryPath, submodule);
      console.log("submodulePath: ", submodulePath);
      const gitSubmoduleRepo = gitP(submodulePath);
      snapshot.submodules[submodule] = await this.takeRepoSnapshot(
        gitSubmoduleRepo,
        allow_to_stash
      );
    }

    // Save the state of the main repository
    snapshot.mainRepo = await this.takeRepoSnapshot(this.git, allow_to_stash);

    // Save the snapshot to this.snapshot
    this.snapshot = snapshot;
    this.hasChanges = this.isRepoDirty(this.git);

    console.log("snapshot: ", snapshot);
  }

  /**
   * Undoes a repository snapshot by checking out the branch or commit hash and restoring the stash if applicable.
   * @param {GitRepo} gitRepo - The Git repository object.
   * @param {Object} snapshot - The snapshot object containing information about the snapshot.
   * @param {string} [snapshot.branch] - The branch to check out.
   * @param {string} [snapshot.commitHash] - The commit hash to check out.
   * @param {string} [snapshot.stashId] - The stash ID to restore.
   * @returns {Promise<void>} - A promise that resolves when the snapshot is undone.
   */
  async undoRepoSnapshot(gitRepo, snapshot) {
    console.log("In undoRepoSnapshot");
    if (snapshot.branch) {
      await gitRepo.checkout(snapshot.branch);
    } else {
      await gitRepo.checkout(snapshot.commitHash);
    }
    if (snapshot.stashId) {
      // Restore the stash
      await gitRepo.stash(["pop", snapshot.stashId]);
    }
  }

  /**
   * Restores the repository to the state saved in the snapshot.
   * Iterates over all submodules and restores the commit hash and branch name if they exist.
   * If a submodule had uncommitted changes, restores the stash.
   * Does the same for the main repository.
   * @throws {Error} If VersionFinder is not initialized.
   */
  async restoreRepoSnapshot() {
    if (!this.isInitialized) {
      throw new Error("VersionFinder is not initialized.");
    }
    console.log("In restoreRepoSnapshot");
    if (this.snapshot) {
      console.log("In restoreRepoSnapshot: snapshot");
      console.log("snapshot: ", this.snapshot);
      /**
       * Restore the repository to the state saved in the snapshot.
       * Iterate over all submodules and restore the commit hash and branch name if exists.
       * If a submodule had uncommited changes, restore the stash.
       * Do the same for the main repository.
       */
      const snapshot = this.snapshot;

      // Iterate over all submodules
      for (const submodule of this.submodules) {
        const submodulePath = path.join(this.repositoryPath, submodule);
        const gitRepo = gitP(submodulePath);
        const submoduleSnapshot = snapshot.submodules[submodule];
        await this.undoRepoSnapshot(gitRepo, submoduleSnapshot);
      }

      // Restore the state of the main repository
      const mainRepoSnapshot = snapshot.mainRepo;
      await this.undoRepoSnapshot(this.git, mainRepoSnapshot);

      // Clear the snapshot
      this.snapshot = null;
      this.hasChanges = true;
    }
  }

  /**
   * Checks if a commit SHA is valid for a given branch and submodule.
   * @param {string} commitSha - The commit SHA to check.
   * @param {string} branch - The branch name.
   * @param {string} submodule - The submodule name. Optional.
   * @returns {Promise<boolean>} - A promise that resolves to true if the commit SHA is valid, false otherwise.
   */
  async isValidCommitSha(commitSha, branch, submodule) {
    if (!this.isInitialized) {
      throw new Error("VersionFinder is not initialized.");
    }
    if (this.hasChanges) {
      throw new Error(
        "The repository has uncommitted changes. Please commit or discard the changes before proceeding."
      );
    }
    await this.saveRepoSnapshot();
    await this.git.checkout(branch);
    await this.git.pull();
    await this.git.subModule(["update", "--init"]);
    if (submodule) {
      await gitP(path.join(this.repositoryPath, submodule)).show([commitSha]);
    } else {
      await this.git.show([commitSha]);
    }
    await this.restoreRepoSnapshot();
    return true;
  }

  /**
   * Checks if a submodule is an ancestor of a target commit hash.
   *
   * @param {string} submodulePath - The path to the submodule.
   * @param {string} target_commit_hash - The target commit hash.
   * @param {string} submodulePointer - The submodule pointer.
   * @returns {boolean} - Returns true if the submodule is an ancestor, false otherwise.
   * @throws {Error} - Throws an error if VersionFinder is not initialized.
   */
  async checkAncestor(submodulePath, target_commit_hash, submodulePointer) {
    if (!this.isInitialized) {
      throw new Error("VersionFinder is not initialized.");
    }
    let isAncestor = true;
    try {
      const command_string = `cd ${submodulePath} && git merge-base --is-ancestor ${target_commit_hash} ${submodulePointer}`;
      const { stderr } = await execPromise(command_string);
      if (stderr) {
        isAncestor = false;
        console.error("stderr:", stderr);
      }
    } catch (error) {
      if (1 === error.code) {
        isAncestor = false;
      } else {
        throw error;
      }
    }
    return isAncestor;
  }

  /**
   * Gets the first commit SHA for a given branch and submodule.
   * @param {string} branch - The branch name.
   * @param {string} submodule - The submodule name. Optional.
   * @returns {Promise<string>} - A promise that resolves to the first commit SHA.
   */
  async getFirstCommitSha(target_commit_hash, branch, submodule) {
    if (!this.isInitialized) {
      throw new Error("VersionFinder is not initialized.");
    }
    if (this.hasChanges) {
      throw new Error(
        "The repository has uncommitted changes. Please commit or discard the changes before proceeding."
      );
    }
    try {
      await this.saveRepoSnapshot();
      console.log(
        "In getFirstCommitSha: target_commit_hash=",
        target_commit_hash,
        "branch=",
        branch,
        "submodule=",
        submodule
      );
      await this.git.checkout(branch);
      await this.git.pull();
      await this.git.subModule(["update", "--init"]);
      if (submodule) {
        let logs = await this.git.log({
          file: submodule,
          format: { hash: "%H", message: "%s" },
        });
        // Convert logs to an array if not already
        logs = logs.all || [];

        // Binary search function
        const binarySearch = async (logs) => {
          let low = 0;
          let high = logs.length - 1;
          const submodulePath = path.join(this.repositoryPath, submodule);
          console.log("submodulePath: ", submodulePath);
          console.log("submodule: ", submodule);
          while (low <= high) {
            const mid = Math.floor((low + high) / 2);
            const log = logs[mid];
            const lsTreeOutput = await this.git.raw([
              "ls-tree",
              log.hash,
              submodule,
            ]);
            const match = lsTreeOutput.match(/\b[0-9a-f]{40}\b/);
            if (!match) continue; // Skip if no match found
            const submodulePointer = match[0];
            const isAncestor = await this.checkAncestor(
              submodulePath,
              target_commit_hash,
              submodulePointer
            );
            if (isAncestor) {
              // If isAncestor, search left (earlier commits)
              high = mid - 1;
            } else {
              // If not ancestor, search right (later commits)
              low = mid + 1;
            }
          }
          // Return the log at the boundary if it's an ancestor, else null
          if (low < logs.length) {
            const lsTreeOutput = await this.git.raw([
              "ls-tree",
              logs[low].hash,
              submodule,
            ]);
            const match = lsTreeOutput.match(/\b[0-9a-f]{40}\b/);
            if (match) {
              const submodulePointer = match[0];
              const isAncestor = await this.checkAncestor(
                submodulePath,
                target_commit_hash,
                submodulePointer
              );
              if (isAncestor) {
                return logs[low];
              }
            }
          }
          return null;
        };

        return await binarySearch(logs.reverse()); // Reverse logs to have them in chronological order for binary search
      } else {
        // Handle the case where there's no submodule
        let commitDetails = await this.git.raw([
          "show",
          "--no-patch",
          '--format={"hash": "%H", "message": "%s"}',
          target_commit_hash,
        ]);
        commitDetails = JSON.parse(commitDetails);
        return commitDetails;
      }
    } catch (error) {
      console.error("Error fetching first commit SHA.");
      console.error(error);
      throw error;
    } finally {
      await this.restoreRepoSnapshot();
    }
  }

  /**
   * Gets the logs for a given branch and submodule.
   * @param {string} branch - The branch name.
   * @param {string} submodule - The submodule name. Optional.
   * @returns {Promise<Object[]>} - A promise that resolves to an array of log objects.
   */
  async getLogs(branch, submodule = null, commit_hash = "HEAD") {
    if (!this.isInitialized) {
      throw new Error("VersionFinder is not initialized.");
    }
    if (this.hasChanges) {
      throw new Error(
        "The repository has uncommitted changes. Please commit or discard the changes before proceeding."
      );
    }
    try {
      await this.saveRepoSnapshot();
      console.log("branch: ", branch);
      console.log("submodule: ", submodule);
      await this.git.checkout(branch);
      // await this.git.pull();
      await this.git.subModule(["update", "--init"]);
      let logs;
      let gitRepo = this.git;
      if (submodule) {
        gitRepo = gitP(path.join(this.repositoryPath, submodule));
      }
      logs = await gitRepo.log({ from: commit_hash });
      // Append the commit_hash to the logs
      const commit_hash_log = await gitRepo.raw([
        "show",
        "--no-patch",
        '--format={ "hash": "%H", "message": "%s"}',
        commit_hash,
      ]);
      logs.all.push(JSON.parse(commit_hash_log));
      return logs.all;
    } catch (error) {
      console.error("Error fetching logs.");
      console.error(error);
      throw error;
    } finally {
      await this.restoreRepoSnapshot();
    }
  }

  /**
   * Gets the first commit SHA that contains a version in the commit message for a given branch and submodule.
   * @param {string} commitSHA - The commit SHA to start searching from.
   * @param {string} branch - The branch name.
   * @param {string} submodule - The submodule name. Optional.
   * @returns {Promise<string|null>} - A promise that resolves to the first commit SHA that contains a version, or null if no such commit is found.
   */
  async getFirstCommitWithVersion(commitSHA, branch, submodule) {
    console.log("In getFirstCommitWithVersion");
    console.log("commitSHA: ", commitSHA);
    console.log("branch: ", branch);
    console.log("submodule: ", submodule);
    if (this.hasChanges) {
      throw new Error(
        "The repository has uncommitted changes. Please commit or discard the changes before proceeding."
      );
    }
    try {
      const logs = await this.getLogs(branch, null, commitSHA);
      console.log("logs: ", logs);
      for (const log of logs.reverse()) {
        if (log.message.match(this.searchPatternRegex)) {
          return log;
        }
      }
      return null;
    } catch (error) {
      console.error(error);
      console.error("Error fetching first commit with version.");
      throw error;
    }
  }
}

module.exports = { VersionFinder };
